vms/subnet.go

```diff
@@ -0,0 +1,364 @@
+	// Create blockchain
+	genesis, err := os.ReadFile(vm.Genesis)
+	if err != nil {
+		return ids.Empty, fmt.Errorf("could not read genesis file (%s): %w", vm.Genesis, err)
+	}
+	txID, err := args.txPChainClient.CreateBlockchain(
+		args.userPass,
+		[]string{args.fundedAddress},
+		args.fundedAddress,
+		args.rSubnetID,
+		vm.ID,
+		[]string{},
+		vm.Name,
+		genesis,
+	)
+	if err != nil {
+		return ids.Empty, fmt.Errorf("could not create blockchain: %w", err)
+	}
+	for {
+		select {
+		case <-ctx.Done():
+			return ids.Empty, ctx.Err()
+		case <-time.After(apiRetryFreq):
+		}
+		status, err := args.txPChainClient.GetTxStatus(txID, true)
+		if err != nil {
+			return ids.Empty, err
+		}
+		if status.Status == platformvm.Committed {
+			args.log.Info("create blockchain tx (%s) accepted", txID)
+			return txID, nil
+		}
+		args.log.Debug("waiting for create blockchain tx (%s) to be accepted", txID)
+	}
+}
+
+// finalizeBlockchain is a checking function. It ensures that the given nodes
+// are validating the blockchain, and that all nodes have the VM bootstrapped.
+// If all is ok, it prints the endpoints to STDOUT, otherwise it returns an error.
+func finalizeBlockchain(ctx context.Context, log logging.Logger, allNodes map[string]node.Node, blockchainID ids.ID) error {
+	if err := ensureValidating(ctx, log, allNodes, blockchainID); err != nil {
+		return fmt.Errorf("error checking all nodes are validating the blockchain: %w", err)
+	}
+	if err := ensureBootstrapped(ctx, log, allNodes, blockchainID); err != nil {
+		return fmt.Errorf("error checking blockchain is bootstrapped: %w", err)
+	}
+	// Print endpoints where VM is accessible
+	log.Info("Custom VM endpoints now accessible at:")
+	for _, n := range allNodes {
+		log.Info("%s: %s:%d/ext/bc/%s", n.GetNodeID(), n.GetURL(), n.GetAPIPort(), blockchainID.String())
+	}
+	return nil
+}
+
+// ensureValidating returns an error if not all of the nodes are validating this
+// blockchain or if waiting for nodes to confirm validation status times out.
+func ensureValidating(tctx context.Context, log logging.Logger, allNodes map[string]node.Node, blockchainID ids.ID) error {
+	statusCheckTimeout := longTimeout
+	// Ensure all nodes are validating subnet
+	g, ctx := errgroup.WithContext(tctx)
+	for _, node := range allNodes {
+		node := node
+		g.Go(func() error {
+			nodeID := node.GetNodeID().PrefixedString(constants.NodeIDPrefix)
+			nClient := node.GetAPIClient().PChainAPI()
+			for {
+				select {
+				case <-ctx.Done():
+					return ctx.Err()
+				case <-time.After(statusCheckTimeout):
+				}
+				status, err := nClient.GetBlockchainStatus(blockchainID.String())
+				if err != nil {
+					return fmt.Errorf("error querying blockchain status: %w", err)
+				}
+				if status == platformvm.Validating {
+					// after the first acceptance, next nodes probably don't need to check that long anymore
+					statusCheckTimeout = apiRetryFreq
```

Comment:
why do we use 2 different wait durations?

Reply:
Because until the nodes are validating there is a longer wait. Until the first node has been recognized to validate though, the other nodes should get validating probably be faster (i.e. at the same time actually), so we don't need long intervals anymore.